File: src/main/java/org/jukeboxmc/command/internal/WhitelistCommand.java
package org.jukeboxmc.command.internal;

import org.cloudburstmc.protocol.bedrock.data.command.CommandParamType;
import org.jukeboxmc.Server;
import org.jukeboxmc.command.Command;
import org.jukeboxmc.command.CommandData;
import org.jukeboxmc.command.CommandParameter;
import org.jukeboxmc.command.CommandSender;
import org.jukeboxmc.command.annotation.Description;
import org.jukeboxmc.command.annotation.Name;
import org.jukeboxmc.command.annotation.Permission;

import java.util.Arrays;
import java.util.List;

/**
 * @author LucGamesYT
 * @version 1.0
 */
@Name ( "whitelist" )
@Description ( "Manages the list of players allowed to use this server" )
@Permission ( "jukeboxmc.command.whitelist" )
public class WhitelistCommand extends Command {

    public WhitelistCommand() {
        super( CommandData.builder()
                .setParameters( new CommandParameter[]{
                                new CommandParameter( "whitelist", Arrays.asList( "on", "off" ), false )
                        },
                        new CommandParameter[]{
                                new CommandParameter( "list", List.of( "list" ), false )
                        },
                        new CommandParameter[]{
                                new CommandParameter( "action", Arrays.asList( "add", "remove" ), false ),
                                new CommandParameter( "player", CommandParamType.TARGET, false )

                        }
                )
                .build() );
    }

    @Override
    public void execute( CommandSender commandSender, String command, String[] args ) {
        if ( args.length == 1 ) {
            if ( args[0].equalsIgnoreCase( "on" ) ) {
                if ( Server.getInstance().hasWhitelist() ) {
                    commandSender.sendMessage( "§cThe whitelist is already activated" );
                    return;
                }
                Server.getInstance().setWhitelist( true );
                commandSender.sendMessage( "The whitelist was activated" );
            } else if ( args[0].equalsIgnoreCase( "off" ) ) {
                if ( !Server.getInstance().hasWhitelist() ) {
                    commandSender.sendMessage( "§cThe whitelist is already disabled" );
                    return;
                }
                Server.getInstance().setWhitelist( false );
                commandSender.sendMessage( "The whitelist was deactivated" );
            } else if ( args[0].equalsIgnoreCase( "list" ) ){
                StringBuilder builder = new StringBuilder();
                for ( String playerName : Server.getInstance().getWhitelist() ) {
                    builder.append( playerName ).append( ", " );
                }
                builder.setLength( builder.length() - 2 );
                commandSender.sendMessage( "The following players are on the whitelist: " + builder );
            } else {
                this.sendUsage( commandSender );
            }
        } else if ( args.length == 2 ) {
            if ( args[0].equalsIgnoreCase( "add" ) ) {
                String playerName = args[1];
                if ( Server.getInstance().isPlayerOnWhitelist( playerName ) ) {
                    commandSender.sendMessage( "§c" + playerName + " is already on the whitelist" );
                    return;
                }
                Server.getInstance().addPlayerToWhitelist( playerName );
                commandSender.sendMessage( playerName + " has been added to the whitelist" );
            } else if ( args[0].equalsIgnoreCase( "remove" ) ) {
                String playerName = args[1];
                if ( !Server.getInstance().isPlayerOnWhitelist( playerName ) ) {
                    commandSender.sendMessage( "§c" + playerName + " is not on the whitelist" );
                    return;
                }
                Server.getInstance().removePlayerFromWhitelist( playerName );
                commandSender.sendMessage( playerName + " was removed from the whitelist" );
            } else {
                this.sendUsage( commandSender );
            }
        } else {
            this.sendUsage( commandSender );
        }
    }

    private void sendUsage( CommandSender commandSender ) {
        commandSender.sendMessage( "Usage:" );
        commandSender.sendMessage( "- /whitelist <add|remove> <player>" );
        commandSender.sendMessage( "- /whitelist <on|off>" );
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/command/internal/WhitelistCommand.kt
package org.jukeboxmc.server.command.internal

import org.jukeboxmc.api.command.Command
import org.jukeboxmc.api.command.CommandSender
import org.jukeboxmc.api.command.ParameterType
import org.jukeboxmc.api.command.annotation.*
import org.jukeboxmc.server.JukeboxServer

@Name("whitelist")
@Description("Manages the list of players allowed to use this server")
@Permission("jukeboxmc.command.whitelist")
@Parameters(parameter = [
    Parameter("whitelist", enumValues = arrayOf("on", "off"), optional = false)
])
@Parameters(parameter = [
    Parameter("list", enumValues = arrayOf("list"), optional = false)
])
@Parameters(parameter = [
    Parameter("action", enumValues = arrayOf("add", "remove"), optional = false),
    Parameter("player", ParameterType.TARGET , optional = false)
])
class WhitelistCommand : Command {

    override fun execute(commandSender: CommandSender, command: String, args: Array<String>) {
        if (args.size == 1) {
            if (args[0].equals("on", ignoreCase = true)) {
                if (JukeboxServer.getInstance().isWhitelist()) {
                    commandSender.sendMessage("§cThe whitelist is already activated")
                    return
                }
                JukeboxServer.getInstance().setWhitelist(true)
                commandSender.sendMessage("The whitelist was activated")
            } else if (args[0].equals("off", ignoreCase = true)) {
                if (!JukeboxServer.getInstance().isWhitelist()) {
                    commandSender.sendMessage("§cThe whitelist is already disabled")
                    return
                }
                JukeboxServer.getInstance().setWhitelist(false)
                commandSender.sendMessage("The whitelist was deactivated")
            } else if (args[0].equals("list", ignoreCase = true)) {
                val builder = StringBuilder()
                for (playerName in JukeboxServer.getInstance().getWhitelist().getPlayers()) {
                    builder.append(playerName).append(", ")
                }
                builder.setLength(builder.length - 2)
                commandSender.sendMessage("The following players are on the whitelist: $builder")
            } else {
                this.sendUsage(commandSender)
            }
        } else if (args.size == 2) {
            if (args[0].equals("add", ignoreCase = true)) {
                val playerName = args[1]
                if (JukeboxServer.getInstance().getWhitelist().isWhitelisted(playerName)) {
                    commandSender.sendMessage("§c$playerName is already on the whitelist")
                    return
                }
                JukeboxServer.getInstance().getWhitelist().add(playerName)
                commandSender.sendMessage("$playerName has been added to the whitelist")
            } else if (args[0].equals("remove", ignoreCase = true)) {
                val playerName = args[1]
                if (!JukeboxServer.getInstance().getWhitelist().isWhitelisted(playerName)) {
                    commandSender.sendMessage("§c$playerName is not on the whitelist")
                    return
                }
                JukeboxServer.getInstance().getWhitelist().remove(playerName)
                commandSender.sendMessage("$playerName was removed from the whitelist")
            } else {
                this.sendUsage(commandSender)
            }
        } else {
            this.sendUsage(commandSender)
        }
    }

    private fun sendUsage(commandSender: CommandSender) {
        commandSender.sendMessage("Usage:")
        commandSender.sendMessage("- /whitelist <add|remove> <player>")
        commandSender.sendMessage("- /whitelist <on|off>")
    }
}